File: Tests/LiteralTests.cs
using System;
using System.Text.Patterns;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests {
	[TestClass]
	public class LiteralTests {
		[TestMethod]
		public void Constructor() {
			Pattern _ = "Hello";
		}

		[TestMethod]
		public void Consume() {
			Pattern Hello = "Hello";
			Pattern Space = ' ';
			Pattern World = "World";
			Result Result;

			Result = Hello.Consume("Hello World!");
			ResultAssert.Captures("Hello", Result);

			Source Source = new Source("Hello World");

			Result = Hello.Consume(ref Source);
			ResultAssert.Captures("Hello", Result);

			Result = Space.Consume(ref Source);
			ResultAssert.Captures(" ", Result);

			Result = World.Consume(ref Source);
			ResultAssert.Captures("World", Result);

			Result = Hello.Consume("Hell");
			ResultAssert.Fails(Result);

			Result = Hello.Consume("Bacon");
			ResultAssert.Fails(Result);

			Pattern HelloWorld = Hello & Space & World; //This isn't a normal concatenator, but rather, should be optimized into a single StringLiteral

			Result = HelloWorld.Consume("HELLO WORLD!");
			ResultAssert.Captures("HELLO WORLD", Result);

			Result = HelloWorld.Consume("Hello World!");
			ResultAssert.Captures("Hello World", Result);

			Result = HelloWorld.Consume("hello world!");
			ResultAssert.Captures("hello world", Result);
		}

		[TestMethod]
		public void Equal() {
			Pattern Hello = "Hello";
			Pattern Space = ' ';
			Pattern World = "World";

			Assert.IsTrue(Hello.Equals("Hello"));
			Assert.IsFalse(Hello.Equals("Hello "));
			Assert.IsFalse(Hello.Equals("Hell"));
			Assert.IsFalse(Hello.Equals("World"));

			Assert.IsTrue(Space.Equals(" "));
			Assert.IsFalse(Space.Equals("  "));
			Assert.IsFalse(Space.Equals("a"));

			Pattern HelloWorld = Hello & Space & World; //This isn't a normal concatenator, but rather, should be optimized into a single StringLiteral

			Assert.IsTrue(HelloWorld.Equals("Hello World"));
			Assert.IsFalse(HelloWorld.Equals("Hello"));
			Assert.IsFalse(HelloWorld.Equals("Hello Bacon"));
		}
	}
}

File: Patterns.Tests/LiteralTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type LiteralTests() =

    [<TestMethod>]
    member _.``constructor`` () =
        let _charLiteral = p 'h'
        let _stringLiteral = p "hello"
        ()

    [<TestMethod>]
    member _.``consume string`` () =
        let hello = p"Hello"
        
        ResultAssert.Captures("Hello", hello.Consume("Hello World!"))

        ResultAssert.Fails(hello.Consume("Hell"))

        ResultAssert.Fails(hello.Consume("Bacon"))

    [<TestMethod>]
    member _.``consume source`` () =
        let hello = p"Hello"
        let space = p ' '
        let world = p"World"

        let mutable source = Source("Hello World")

        ResultAssert.Captures("Hello", hello.Consume(&source))

        ResultAssert.Captures(" ", space.Consume(&source))

        ResultAssert.Captures("World", world.Consume(&source))

        source <- Source("Hello World")
        let helloWorld = hello >> space >> world

        ResultAssert.Captures("Hello World", helloWorld.Consume(&source))

    [<TestMethod>]
    member _.``consume case-insensitive`` () =
        let pattern = "Hello"/=Compare.CaseInsensitive >> ' '/=Compare.CaseInsensitive >> "World"/=Compare.CaseInsensitive
        
        ResultAssert.Captures("HELLO WORLD", pattern.Consume("HELLO WORLD"))
        ResultAssert.Captures("hello world", pattern.Consume("hello world"))

    [<TestMethod>]
    member _.``neglect string`` () =
        let pattern = not "Hello"
        ResultAssert.Fails(pattern.Consume("Hello"))
        ResultAssert.Fails(pattern.Consume("Hello!"))
        ResultAssert.Fails(pattern.Consume("Hello."))
        ResultAssert.Captures("Oh no", pattern.Consume("Oh no!"))
        ResultAssert.Captures("Oh no", pattern.Consume("Oh no?"))